From 6532f8806ba5d161c95b48da314effc71ad74720 Mon Sep 17 00:00:00 2001
From: Shriram K <shriram.k@arm.com>
Date: Tue, 25 Oct 2022 13:08:20 +0530
Subject: [PATCH] [TEMP] drivers/arm: add function to trigger mhuv3
 doorbell

Add a new function to trigger the doorbell channels in MHUv3. This
function will be used by the RD-Fremont platform when MHUv3 is used
instead of MHUv2.

Upstream-Status: Pending
Signed-off-by: Shriram K <shriram.k@arm.com>
Signed-off-by: Peter Hoyes <Peter.Hoyes@arm.com>
---
 drivers/arm/css/mhu/css_mhu_doorbell.c     | 9 +++++++++
 include/drivers/arm/css/css_mhu_doorbell.h | 5 +++++
 2 files changed, 14 insertions(+)

diff --git a/drivers/arm/css/mhu/css_mhu_doorbell.c b/drivers/arm/css/mhu/css_mhu_doorbell.c
index c51f3b1d7..89c0a12ef 100644
--- a/drivers/arm/css/mhu/css_mhu_doorbell.c
+++ b/drivers/arm/css/mhu/css_mhu_doorbell.c
@@ -38,3 +38,12 @@ void mhuv2_ring_doorbell(struct scmi_channel_plat_info *plat_info)
 
 	return;
 }
+
+void mhuv3_ring_doorbell(struct scmi_channel_plat_info *plat_info)
+{
+	MHU_RING_DOORBELL(plat_info->db_reg_addr,
+			plat_info->db_modify_mask,
+			plat_info->db_preserve_mask);
+
+	return;
+}
diff --git a/include/drivers/arm/css/css_mhu_doorbell.h b/include/drivers/arm/css/css_mhu_doorbell.h
index 88302fd7b..479113630 100644
--- a/include/drivers/arm/css/css_mhu_doorbell.h
+++ b/include/drivers/arm/css/css_mhu_doorbell.h
@@ -22,6 +22,10 @@
 #define SENDER_REG_STAT(_channel)	(0x20 * (_channel))
 #define SENDER_REG_SET(_channel)	((0x20 * (_channel)) + 0xC)
 
+#define MHU3_PBX_PDBCW_PAGE_OFFSET      UINT32_C(0x00001000)
+#define MHU3_SENDER_REG_STAT(_channel)	(MHU3_PBX_PDBCW_PAGE_OFFSET + (0x20 * (_channel)))
+#define MHU3_SENDER_REG_SET(_channel)	(MHU3_PBX_PDBCW_PAGE_OFFSET + ((0x20 * (_channel)) + 0xC))
+
 /* Helper macro to ring doorbell */
 #define MHU_RING_DOORBELL(addr, modify_mask, preserve_mask)	do {	\
 		uint32_t db = mmio_read_32(addr) & (preserve_mask);	\
@@ -40,5 +44,6 @@
 struct scmi_channel_plat_info;
 void mhu_ring_doorbell(struct scmi_channel_plat_info *plat_info);
 void mhuv2_ring_doorbell(struct scmi_channel_plat_info *plat_info);
+void mhuv3_ring_doorbell(struct scmi_channel_plat_info *plat_info);
 
 #endif	/* CSS_MHU_DOORBELL_H */
-- 
2.34.1

